### PR TITLE
Add FileScanClient and business logic

### PR DESCRIPTION
**Pull request by SandGrainOne**
## Description
Add FileScanClient to enqueue file scan requests
Add DataService to hold the (very limited) business logic.
Update Program to register new services

## Related Issue(s)
- https://github.com/Altinn/altinn-file-scan/issues/15

### CODE REVIEW

File: src/Storage/Controllers/DataController.cs [hunk 1]
```diff
@@ -287,6 +294,8 @@ public async Task<ActionResult<DataElementList>> GetMany(int instanceOwnerPartyI
 
             DataElement dataElement = await _dataRepository.Create(newData);
             dataElement.SetPlatformSelfLinks(_storageBaseAndHost, instanceOwnerPartyId);
+            
+            await _dataService.PerformFileScan(dataTypeDefinition, dataElement, CancellationToken.None);
```
Comment (acn-sbuad) [inline src/Storage/Controllers/DataController.cs:298]: I think we need this for the PUT endpoint as well, in those cases they PUT on an existing data object

File: src/Storage/appsettings.json [hunk 2]
```diff
@@ -32,5 +32,9 @@
   },
   "PepSettings": {
     "PdpDecisionCachingTimeout": 5
+  },
+  "FileScanQueueSettings": {
+    "FileScanQueueName": "file-scan-inbound",
+    "ConnectionString": "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;QueueEndpoint=http://127.0.0.1:10001/devstoreaccount1"
```
Comment (acn-sbuad) [inline src/Storage/appsettings.json:38]: I don't think we have a task for it, but this connection string should be stored in KV during generation of the storage account. Perhaps Consider agreeing on naming of the property with @tba76 so the secret is only stored once, but used by both storage and the function? given that we are referencing the same resource

Comment (acn-sbuad) [reply src/Storage/appsettings.json:38]: Connection string is persisted in key vault during provisioning as `QueueStorageSettings--ConnectionString`

File: src/Storage/Controllers/DataController.cs (hunk 1 above)
Comment (SandGrainOne) [reply src/Storage/Controllers/DataController.cs]: Added as requested

File: src/Storage/appsettings.json (hunk 2 above)
Comment (acn-sbuad) [reply src/Storage/appsettings.json:38]: My bad. The function secret is defined seperate of what is loaded into key vault, so duplicating settings name from the function won't work. We have two options here
1) Name it `QueueStorageSettings` to match what is is KV today for this storage account
2) Update tf to duplicate the secret with the desired name `FileScanQueueSettings` in KV during provisioning.

Thinking of it, maybe a 3rd as well.. Read the QueueStorageSettings into the `FileScanQueueSettings` class when setting up the configuration in Program.cs. Although this would make it hard to follow the code and find the settings use in kv at a later point I guess

File: test/UnitTest/TestingControllers/DataControllerTests.cs [hunk 3]
```diff
@@ -65,13 +66,34 @@ public async void Post_NewData_Ok()
...
+        }
+
+        [Fact]
+        public async void Post_NewDataThatRequiresFileScan_Ok()
+        {
+            // Arrange
+            string dataPathWithData = $"{_versionPrefix}/instances/1337/bc19107c-508f-48d9-bcd7-54ffec905306/data";
+            HttpContent content = new StringContent("This is a blob file");
+
+            Mock<IFileScanQueueClient> fileScanMock = new Mock<IFileScanQueueClient>();
```
Comment (acn-sbuad) [inline test/UnitTest/TestingControllers/DataControllerTests.cs:79]: any reasoning behind mocking the queueClient rather than the dataService that the controller has a dependency on? Wouldn't it be cleaner to mock the immediate dependency ?

File: src/Storage/Program.cs [hunk 4]
```diff
@@ -242,11 +243,13 @@ void ConfigureServices(IServiceCollection services, IConfiguration config)
     services.AddSingleton<ISasTokenProvider, SasTokenProvider>();
     services.AddSingleton<IKeyVaultClientWrapper, KeyVaultClientWrapper>();
     services.AddSingleton<IPDP, PDPAppSI>();
+    services.AddSingleton<IFileScanQueueClient, FileScanQueueClient>();
```
Comment (acn-sbuad) [inline src/Storage/Program.cs:246]: We might be missing the line to adding the FileScanQueueSettings to the configuration?

File: src/Storage/appsettings.json (hunk 2 above)
Comment (acn-sbuad) [reply src/Storage/appsettings.json:38]: To clarify, the issue isn't the queue name, but the if for the settings section `FileScanQueueSettings` which we will need to override with a connection from kv.

File: test/UnitTest/TestingControllers/DataControllerTests.cs (hunk 3 above)
Comment (SandGrainOne) [reply test/UnitTest/TestingControllers/DataControllerTests.cs:79]: These tests already include a full in memory http service with an almost complete middleware. In my opinion not very clean to begin with. 

Whenever I use the in memory http service I like to mock as little as possible, or as late as possible. This gives them "deeper" integrations.  This is different from unit tests were I like to mock immediate dependencies.

File: src/Storage/Program.cs (hunk 4 above)
Comment (SandGrainOne) [reply src/Storage/Program.cs:246]: Yes, good catch.

**Review by acn-sbuad — APPROVED** [review]
🎈

File: src/Storage/Altinn.Platform.Storage.csproj [hunk 5]
```diff
@@ -9,12 +9,13 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Altinn.Platform.Storage.Interface" Version="3.11.0" />
+    <PackageReference Include="Altinn.Platform.Storage.Interface" Version="3.14.0" />
```
Comment (acn-sbuad) [inline src/Storage/Altinn.Platform.Storage.csproj:12]:
```suggestion
    <PackageReference Include="Altinn.Platform.Storage.Interface" Version="3.15.0" />
```

File: src/Storage/Controllers/DataController.cs [hunk 6]
```diff
@@ -287,6 +296,8 @@ public async Task<ActionResult<DataElementList>> GetMany(int instanceOwnerPartyI
 
             DataElement dataElement = await _dataRepository.Create(newData);
             dataElement.SetPlatformSelfLinks(_storageBaseAndHost, instanceOwnerPartyId);
+
```
Comment (acn-sbuad) [inline src/Storage/Controllers/DataController.cs:299]: Should the logic to add the hash to the data element be included here somewhere ?

File: src/Storage/Program.cs (hunk 4 above)
Comment (b-brodie-digdir) [reply src/Storage/Program.cs:247]: Resolved on line 195?

**Review by b-brodie-digdir — COMMENTED** [review]
Looks good to me.

File: src/Storage/appsettings.json (hunk 2 above)
Comment (SandGrainOne) [reply src/Storage/appsettings.json:38]: I would have preferred option 2, but have now implemented 1.
